src/consumers/manager.rs

```diff
@@ -88,22 +99,85 @@ impl<C> ConsumerManager<C> {
     /// Returns a new consumer manager set up to use the given permit pool. This meant to use a
     /// shared pool of permits with other consumer managers to manage the amount of simultaneous
     /// work, so the Semaphore must be wrapped in an [`Arc`].
-    pub fn new(permit_pool: Arc<Semaphore>, stream: NatsStream) -> ConsumerManager<C> {
-        ConsumerManager {
+    ///
+    /// This function will attempt to populate this itself with all existing consumers. Any errors
+    /// that occur during population will only log and not error out as it is recoverable. Because
+    /// of this, it requires something that can generate the desired worker
+    pub async fn new<W, F>(
+        permit_pool: Arc<Semaphore>,
+        stream: NatsStream,
+        worker_generator: F,
+    ) -> ConsumerManager<C>
+    where
+        W: Worker + Send + Sync + 'static,
+        C: Stream<Item = Result<ScopedMessage<W::Message>, async_nats::Error>>
+            + CreateConsumer<Output = C>
+            + Send
+            + Unpin
+            + 'static,
+        F: WorkerCreator<Output = W>,
+    {
+        let mut manager = ConsumerManager {
             handles: Arc::new(RwLock::new(HashMap::default())),
             permits: permit_pool,
             stream,
             phantom: PhantomData,
-        }
+        };
+
+        let handles: HashMap<String, JoinHandle<WorkResult<()>>> = manager.stream.consumers().filter_map(|res| async {
+            let info = match res {
```

Comment:
I think you could use the fancy `if let Ok... else` syntax here yeah?

Reply:
Yeah, it was mostly a style choice to use match. I think I started with `if let` and didn't like it as much

Reply:
Oh wait, can't capture the error that way